### PR TITLE
move nuget update analysis to native code

### PR DESCRIPTION
**Pull request by brettfo**
The accompanying smoke test PR is:
* dependabot/smoke-tests#211

----

This PR adds a new native command `analyze` that takes the place of the `update_checker` in Ruby.  The purpose of this code is to query the available NuGet package sources to see what updates are available.  With this change we now no longer manually poke any NuGet APIs; we're using "real" NuGet bits for all of it which means we shouldn't have any more issues around not-quite-compliant NuGet feeds.

There is also a test change in Ruby to stub the calls into the native C# tool to instead simply dump a pre-baked JSON response.  This takes several minutes off of the CI time.

----

### Experiments

There is a lot of churn in this PR, so to mitigate some of that, a large bulk of work was placed behind the `nuget_native_analysis` flag, but not all.

Changes that this PR incorporates whether or not the experiment is enabled:
- Workspace discovery (i.e., through `file_parser.rb`) has been updated to always report the full path of dependency files (e.g., `/src/Directory.Build.props` instead of just `Directory.Build.props`.  This is required to avoid confusion during a multi-directory update, because it's possible for two directories to be given, `src/` and `test/`, and if both have a `Directory.Build.props` file, later update stages don't know which file is being referenced.  This change is reflected in the smoke tests.
- Metadata finder has been scaled back to barebones behavior.  Workspace discovery, mentioned above, now reports the package info URL directly, instead of leaving it up to the metadata finder to manually poke the NuGet APIs to find a `.nuspec` file and then extract the repo.  This change is reflected in the smoke tests.

Changes placed behind the experiments flag:
- The file `update_checker.rb` has been modified to check for the experiment flag and dispatch to the new update checker in `native_update_checker/native_update_checker.rb` when appropriate.  If the experiment flag is not set, the old behavior remains.  To disambiguate files and types with similar names, I've prepended `native_` or `Native` to the new ones.

I have tried to keep the changes that check the experiments flag, move the new update checker, and apply the prefix `Native...` to a single commit so that when the experiment is over, that single commit can essentially be reverted (modulo some minor cleanup) to enable the new behavior.

I have verified the old and new behavior is invoked as appropriate by running an update job both with and without the experiments flag in the `job.yml` file.

----

I timed some update operations and the new behavior appears to be within a few seconds of the old/existing behavior with quite an overlap, so the end result is probably a wash.

Existing updater, 2 runs: 5:23 and 6:16
New updater, 2 runs: 5:51, 6:29

### CODE REVIEW

**Review by edgarrs — APPROVED** [review]

File: .github/workflows/smoke.yml [hunk 1]
```diff
@@ -16,6 +16,7 @@ concurrency:
 
 env:
   GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  SMOKE_TEST_BRANCH: main
```
Comment (jeffwidman) [inline .github/workflows/smoke.yml:19]: nice! Making life simple down the road.

**Review by jeffwidman — APPROVED** [review]
I mostly skimmed it, and trusting your team to own this section of code. What parts I did poke at deeper looked good to me.

> I have tried to keep the changes that check the experiments flag, move the new update checker, and apply the prefix `Native`... to a single commit so that when the experiment is over, that single commit can essentially be reverted (modulo some minor cleanup) to enable the new behavior.

😍

**Comment by jeffwidman** [comment]
Merging via a merge commit to preserve history... I find the "fix lint" commits ugly, but I don't want to squash down and lose the niceness where you'd put most of the feature-flag changes behind a single commit.

File: .github/workflows/smoke.yml (hunk 1 above)
Comment (brettfo) [reply .github/workflows/smoke.yml:19]: I've had to re-figure this out too many times.

**Pull request merged**